catalog/core/catalog-core-encryption/src/main/java/org/codice/ddf/catalog/content/encryption/ContentEncryptorDecryptor.java

```diff
@@ -0,0 +1,317 @@
+  private static final String KEYSTORE_PATH =
+      AccessController.doPrivileged((PrivilegedAction<String>) SecurityConstants::getKeystorePath);
+  private static final String KEYSTORE_PASSWORD =
+      AccessController.doPrivileged(
+          (PrivilegedAction<String>) SecurityConstants::getContentEncryptionKeyPassword);
+  private static final String DDF_HOME =
+      AccessController.doPrivileged(
+          (PrivilegedAction<String>) () -> System.getProperty("ddf.home"));
+  private static final String AES = "AES";
+  private static final String CIPHER_TYPE = "AES/GCM/NoPadding";
+  private static final String ENABLE_DEBUG_LOGGING_MSG =
+      " Enable debug logging for more information.";
+  private static final int KEY_LENGTH = 128;
+  private static final int KEY_BYTE_SIZE = KEY_LENGTH / 8;
+  private static final int IV_BYTE_SIZE = 12;
+  private static final int AUTH_TAG_LENGTH = 128;
+  private static final int AUTH_TAG_BYTE_SIZE = AUTH_TAG_LENGTH / 8;
+  private static final int EXTRA_BYTES = IV_BYTE_SIZE + AUTH_TAG_BYTE_SIZE;
+
+  private static SecureRandom secureRandom = new SecureRandom();
+
+  @VisibleForTesting Cipher cipher;
+  @VisibleForTesting KeyStore keyStore;
+  private SecretKey contentEncryptionKey;
+  private boolean isContentEncryptionEnabled;
+
+  public ContentEncryptorDecryptor() {
+    isContentEncryptionEnabled = SecurityConstants.getContentEncryptionEnabled();
+    if (!isContentEncryptionEnabled) {
+      return;
+    }
+    initKeystore();
+    initKey();
+    initCipher();
+  }
+
+  @VisibleForTesting
+  public ContentEncryptorDecryptor(SecretKey secretKey) {
+    isContentEncryptionEnabled = SecurityConstants.getContentEncryptionEnabled();
+    contentEncryptionKey = secretKey;
+    initCipher();
+  }
+
+  @VisibleForTesting
+  ContentEncryptorDecryptor(boolean initialize) {
+    isContentEncryptionEnabled = SecurityConstants.getContentEncryptionEnabled();
+    if (initialize) {
+      initKeystore();
+      initKey();
+      initCipher();
+    }
+  }
+
+  @Override
+  public InputStream encrypt(InputStream inputStream) throws EncryptorDecryptorException {
+    if (!isContentEncryptionEnabled) {
+      return inputStream;
+    }
+
+    readyCipherForEncryption();
+    return doEncryptionWithIV(cipher.getIV(), inputStream);
+  }
+
+  @Override
+  public InputStream decrypt(InputStream inputStream) throws EncryptorDecryptorException {
+    if (!isContentEncryptionEnabled) {
+      return inputStream;
+    }
+
+    PushbackInputStream wrappedInputStream = new PushbackInputStream(inputStream);
+
+    byte[] iv = new byte[IV_BYTE_SIZE];
+    try {
+      int bytesRead = wrappedInputStream.read(iv, 0, IV_BYTE_SIZE);
+      if (bytesRead != IV_BYTE_SIZE) {
+        throw new IOException(
+            String.format(
+                "Bytes read (%d) does not match IV byte size (%d)", bytesRead, IV_BYTE_SIZE));
+      }
+      wrappedInputStream.unread(IV_BYTE_SIZE - 1);
```

Comment:
❗️ This isn't going to unread the last read byte (except in very rare circumstances). This is going to "unread" (in other words, pushback) the byte value of `IV_BYTE_SIZE - 1` to the end of the stream. This needs to be
`wrappedInputStream.unread(iv[IV_BYTE_SIZE - 1])`

That's the last byte read from the stream, pulled from the `iv` buffer into which you'd read it on line 129.

Reply:
Great catch! But after looking at it, looks like I don't even need the `PushbackInputStream`.

Pretty much, we are appending the IV bytes to the encrypted data, that way we know the IV for any content stored.

However, we want to decrypt the content *only* (not the IV bytes), so when we read the IV from the resource's input stream, we want the input stream to stay at that spot in reading, so when we pass it to the `cipher` it starts reading from the beginning of the encrypted data and not the beginning of the IV.